### PR TITLE
Empty grad fix

### PR DESCRIPTION
**Pull request by jeffra**
This fixes a case where there is an imbalance between empty gradients across ranks. Example: rank 0 has 2 parameters with gradients and rank 1 has 1 parameter with gradients and 1 parameter where grads are None. This caused the all-reduce to be imbalanced in size since we we're previously ignoring all grads that were None. Instead now we pad all None gradients with zero tensors. Unfortunately this imbalance did not cause a crash it would cause hanging issues in the first place that tried to access the reduced gradient data itself.

### CODE REVIEW

File: deepspeed/pt/deepspeed_light.py [hunk 1]
```diff
@@ -979,7 +979,12 @@ def allreduce_no_retain(self, bucket, numel_per_bucket=500000000):
     def buffered_allreduce_fallback(self, grads=None, elements_per_buffer=500000000):
         grads = []
         for param_name, param in self.module.named_parameters():
-            if param.grad is not None:
+            if param.grad is None:
```
Comment (frankseide) [inline deepspeed/pt/deepspeed_light.py:982]: May I suggest to add a comment here to explain why this change was necessary? It may be useful for future readers of the code who may not have had the experience we ran into. Thanks!

Comment (samyam) [inline deepspeed/pt/deepspeed_light.py:982]: I wonder if averaging these gradients with world_size still makes sense when only some of the gradients are non-zero and others are none. Maybe something to just add a comment here ?

File: tests/unit/test_fp16.py [hunk 2]
```diff
@@ -97,13 +98,14 @@ def test_lamb_fp16_empty_grad(tmpdir):
...
-    model = SimpleModel(hidden_dim, empty_grad=True)
+    model = SimpleModel(hidden_dim, empty_grad=True, rank=args.local_rank)
 
-    @distributed_test(world_size=[1])
+    @distributed_test(world_size=[2])
     def _test_lamb_fp16_empty_grad(args, model, hidden_dim):
         model, _, _,_ = deepspeed.initialize(args=args,
                                              model=model,
                                              model_parameters=model.parameters())
+        model.module.rank = args.local_rank
```
Comment (tjruwase) [inline tests/unit/test_fp16.py:108]: Why is this needed?

Comment (jeffra) [reply tests/unit/test_fp16.py:108]: good catch, it's not needed but forgot to remove it in my commit.
